maliput/maliput_sim
# Foxglove visualization

# 🎉 New feature

This PR is a proposal and the goal of this is to generate debate in order to consolidate the visualization API

## Summary
Note: During last F2F talks the requirement for visualizing it in real time has been relaxed decoupling the simulation from the visualization.

 - After each simulation step the current simulation state is stored. The implementation of this is rather simple by now: This might need extra work and debate on how better save the values. As the simulation state is expected to be read afterward for some visualization tool or plotting tool, I would say that using some human-readable format, like json might be useful.
 - A FoxgloveVisualization implementation is provided: 
   - Context: Foxglove Studio allows visualization of live data via websocket([Foxglove websocket](https://foxglove.dev/docs/studio/connection/custom)) data and from a [MCAP](https://foxglove.dev/docs/studio/connection/mcap) file previously recorded(info dumped as protobuf messages). Adapting the data to Foxglove data is powerful in terms of features for the visualization.
   - This PR allows dumping the simulation information into a MCAP file that can be later on loaded using Floxglove Studion (both in the browser or desktop app)
     - The message being recorded in the MCAP file is the [SceneUpdate](https://foxglove.dev/docs/studio/messages/scene-update) in order to be loaded by the 3D panel. However other topics could be published so as to use other tools from Foxglove.



## Test it

After building and sourcing:
1. Execute example I am adding in this PR:
   `python3 src/maliput_sim/examples/rail_car_agent_foxglove_viz.py`
   This will generate a `simulation.mcap` file 
2. Load the `simulation.mcap` file using Foxglove online: https://studio.foxglove.dev/
   ![image](https://github.com/maliput/maliput_sim/assets/53065142/de145a62-823e-4130-b60e-0338089f5fee)

3. Play around with the interface. Note the layout can be modified and saved.

https://github.com/maliput/maliput_sim/assets/53065142/d8be1ea9-1c23-4206-bbe6-4e3a6f24bbd9
 - Note that in the second part of the video I am pausing the sim and manually stepping forward.



## Checklist
- [ ] Consider updating Python bindings (if it affects the public API)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.
